src/server/clients_info/client_visibility.rs

```diff
@@ -0,0 +1,284 @@
+use bevy::{
+    prelude::*,
+    utils::{EntityHashMap, EntityHashSet},
+};
+
+use super::VisibilityPolicy;
+
+/// Entity visibility settings for a client.
+pub struct ClientVisibility {
+    filter: VisibilityFilter,
+    entity_state: EntityState,
+}
+
+impl ClientVisibility {
+    /// Creates a new instance based on the preconfigured policy.
+    pub(super) fn new(policy: VisibilityPolicy) -> Self {
+        match policy {
+            VisibilityPolicy::All => Self::with_filter(VisibilityFilter::All {
+                just_connected: true,
+            }),
+            VisibilityPolicy::Blacklist => Self::with_filter(VisibilityFilter::Blacklist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+            VisibilityPolicy::Whitelist => Self::with_filter(VisibilityFilter::Whitelist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+        }
+    }
+
+    /// Creates a new instance with specific filter.
+    fn with_filter(filter: VisibilityFilter) -> Self {
+        Self {
+            filter,
+            entity_state: Default::default(),
+        }
+    }
+
+    /// Resets the filter state to as it was after [`Self::new`].
+    ///
+    /// `entity_state` remains untouched.
+    pub(super) fn clear(&mut self) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => *just_connected = true,
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            }
+            | VisibilityFilter::Whitelist {
+                list,
+                added,
+                removed,
+            } => {
+                list.clear();
+                added.clear();
+                removed.clear();
+            }
+        }
+    }
+
+    /// Marks all entities as not "just added" and clears the removed.
+    ///
+    /// Should be called after each tick.
+    pub(crate) fn update(&mut self) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => *just_connected = false,
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            } => {
+                for entity in removed.drain() {
+                    list.remove(&entity);
+                }
+                added.clear();
```

Comment:
Because for blacklist I store boolean which indicates that the entity is going to be removed from the list. It's needed to optimize the lookups, see https://github.com/lifescapegame/bevy_replicon/pull/174#discussion_r1451828399. Since I don't actually remove from the list right away, I need to remove it here.